Lib/subprocess.py

```diff
@@ -412,10 +426,53 @@ def run(*popenargs, input=None, timeout=None, check=False, **kwargs):
             stdout, stderr = process.communicate()
             raise TimeoutExpired(process.args, timeout, output=stdout,
                                  stderr=stderr)
-        except:
-            process.kill()
-            process.wait()
-            raise
+        except BaseException as ex:
+            try:
+                if cleanup_timeout is None:
+                    # Choose a reasonable default
+                    if timeout:
+                        remaining_timeout = timeout - (_time() - start_time)
+                        cleanup_timeout = max(0.0, remaining_timeout)
+                    elif isinstance(ex, KeyboardInterrupt):
+                        # KeyboardInterrupts are:
+                        # 1. Typically automatically triggered in the child, too.
+                        #    Therefore, the child is likely to be already cleaning up
+                        #    (if we give it the chance).
+                        # 2. Generated interactively by humans at a keyboard.
+                        #    If the child is well-written (exits in response to SIGINT),
+                        #    a 1 second timeout is usually enough time for it to clean up.
+                        #    Othwerwise, 1 second seems like a tolerable delay in response to Ctrl+C,
+                        #    and a resonable price to pay for correct behavior in the well-written case.
+                        cleanup_timeout = 1.0
+                    else:
+                        # The above reasoning does not apply to arbitrary exceptions,
+                        # so kill the child immediately.
+                        cleanup_timeout = 0.0
+
+                if cleanup_timeout == 0.0:
+                    raise # Skip cleanup
+
+                # Close streams to/from child
+                streams = (process.stdin, process.stdout, process.stderr)
+                for f in filter(None, streams):
+                    try:
+                        f.close()
+                    except:
+                        pass  # Ignore EBADF or other errors.
+
+                try:
+                    process.wait(timeout=cleanup_timeout)
+                except:
+                    # Ignore cleanup errors;
+                    # original exception is always re-raised below
+                    pass
+            finally:
+                # Kill the child and re-raise original exception,
+                # regardless of successful/failed cleanup
+                process.kill()
+                process.wait()
+                raise ex
+
```

Comment:
I think we need much simple solution.

    except:
        if kill_timeout is not None:
            process.terminate()
            try:
                process.wait(timeout=kill_timoeut)
             except:
                 pass
        process.kill()
        process.wait()
        raise

Reply:
Regardless of the default timeout value, I do like this proposal to explicitly `terminate()` the child, rather than merely hoping it will just exit gracefully on its own after we close its stdin/stdout.

One open question is whether or not it is bad to send a `SIGTERM` to a process that has (probably) already received a `SIGINT`.  I would imagine that in most cases, it's not a huge problem.  (If the child is a Python program, then the `SIGTERM` is ignored by default, for example.)  In any case, sending `SIGINT`+`SIGTERM` is still better than defaulting to an immediate `SIGKILL`.

Reply:
Regardless of the default timeout value, I do like this proposal to explicitly `terminate()` the child, rather than merely hoping it will just exit gracefully on its own after we close its stdin/stdout.

One open question is whether or not it is bad to send a `SIGTERM` to a process that has (probably) already received a `SIGINT`.

For example, if the child is a Python program, then the default behavior of `SIGTERM` is to end the process without any cleanup.  If it has already received a `SIGINT`, then the process will likely end *during* cleanup.

Reply:
From the perspective of the child process, a TERM is at worst equivalent to a KILL, and possibly much more polite, since it _can_ be caught.  So, if we are committed to killing the child process, I'd encourage a TERM first.

(It's tempting to full-on generalize this and allow something like `timeout_handler=some_func_accepting_the_process_as_arg` or `err_kill=( (0.5, SIGTERM), (0.5, SIGQUIT), (1.0, SIGKILL) )`  Those args are just illustrative, but you get the idea.)

Reply:
>I'd encourage a TERM first.

👍 

>It's tempting to full-on generalize this

Agree with the sentiment.  I also completely agree with what you implied by using the word "tempting" (instead of, say, "justified" 😄 ).  The `run()` API is meant for easy, general usage, with reasonable default behavior.  If a programmer cares about nuanced error-handling, she should just use the `Popen` API.